### PR TITLE
SC: prevent incremental vote data in Bridge contract.

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes
This PR added the routine removing old vote data to prevent increasing vote data.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: contracts/bridge/BridgeOperator.sol [hunk 1]
```diff
@@ -51,29 +56,40 @@ contract BridgeOperator is Ownable {
...
-        if (!votes[_voteKey][msg.sender]) {
-            votes[_voteKey][msg.sender] = true;
-            require(votesCounts[_voteKey] < votesCounts[_voteKey] + 1, "votesCounts overflow");
-            votesCounts[_voteKey]++;
+        VotesData storage vote = votes[uint8(_voteType)][_nonce];
+
+        if (!vote.voted[_voteKey][msg.sender]) {
+            vote.voted[_voteKey][msg.sender] = true;
+
+            require(vote.voteCounts[_voteKey] < vote.voteCounts[_voteKey] + 1, "voteCounts overflow");
```
Comment (jeonghonah) [inline contracts/bridge/BridgeOperator.sol]: operatorThresholds is uint8. Is this condition really possible?

File: contracts/bridge/BridgeOperator.sol [hunk 2]
```diff
@@ -51,29 +56,40 @@ contract BridgeOperator is Ownable {
...
         internal
         returns(bool)
     {
         require(!closedValueTransferVotes[_requestNonce], "closed vote");
 
-        if (voteCommon(VoteType.ValueTransfer, keccak256(msg.data))) {
+        bytes32 voteKey = keccak256(msg.data);
+        if (voteCommon(VoteType.ValueTransfer, _requestNonce, voteKey)) {
             closedValueTransferVotes[_requestNonce] = true;
             return true;
```
Comment (jeonghonah) [inline contracts/bridge/BridgeOperator.sol]: Do you have any reason not to call removeVoteData here?

File: contracts/bridge/BridgeTransferCommon.sol [hunk 3]
```diff
@@ -127,11 +127,17 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
...
             upperHandleNonce = _requestedNonce;
         }
-        for (i = lowerHandleNonce; i <= upperHandleNonce && handleNoncesToBlockNums[i] > 0; i++) { }
-        lowerHandleNonce = i;
-        if (i != 0) {
-            recoveryBlockNumber = handleNoncesToBlockNums[i-1];
+        for (i = lowerHandleNonce; i <= upperHandleNonce && handleNoncesToBlockNums[i] > 0; i++) {
+            recoveryBlockNumber = handleNoncesToBlockNums[i];
+            delete handleNoncesToBlockNums[i];
+            delete closedValueTransferVotes[i];
```
Comment (jeonghonah) [inline contracts/bridge/BridgeTransferCommon.sol]: If we delete closedValueTransferVotes[i] here, how can we check and prevent duplicated voting?

File: contracts/bridge/BridgeTransferCommon.sol [hunk 4]
```diff
@@ -127,11 +127,17 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
...
+            removeVoteData(VoteType.ValueTransfer, i);
```
Comment (jeonghonah) [inline contracts/bridge/BridgeTransferCommon.sol]: How about calling removeVoteData just after success to vote?

File: contracts/bridge/BridgeTransferCommon.sol [hunk 5]
```diff
@@ -127,11 +127,17 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
...
         }
+        lowerHandleNonce = i;
+    }
+
+    function lowerHandleNonceCheck(uint64 _requestedNonce) internal {
+        require(lowerHandleNonce <= _requestedNonce, "removed vote");
```
Comment (jeonghonah) [inline contracts/bridge/BridgeTransferCommon.sol]: lowerHandleNonce is somewhat like a hint for the recovery. However, if there is huge nonce jump, there is gap between lowerHandleNonce and the largest nonce. In that case, can we check the duplicated vote safely?

File: contracts/bridge/BridgeOperator.sol (hunk 2 above)
Comment (KimKyungup) [reply contracts/bridge/BridgeOperator.sol]: Good. As you said, `closedValueTransferVotes` can prevent to re-vote. So `removeVoteData` can be moved here.

File: contracts/bridge/BridgeTransferCommon.sol (hunk 3 above)
Comment (KimKyungup) [reply contracts/bridge/BridgeTransferCommon.sol]: `lowerHandleNonceCheck(_requestedNonce)` can prevent it.

File: contracts/bridge/BridgeTransferCommon.sol (hunk 5 above)
Comment (KimKyungup) [reply contracts/bridge/BridgeTransferCommon.sol]: The nonces in gap between `lowerHandleNonce` and `largest nonce` can be prevented to vote again by `closedValueTransferVotes`.

File: contracts/bridge/BridgeOperator.sol (hunk 1 above)
Comment (KimKyungup) [reply contracts/bridge/BridgeOperator.sol]: Right I removed it. :)

File: contracts/bridge/BridgeTransferCommon.sol [hunk 6]
```diff
@@ -127,11 +127,16 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
...
             upperHandleNonce = _requestedNonce;
         }
-        for (i = lowerHandleNonce; i <= upperHandleNonce && handleNoncesToBlockNums[i] > 0; i++) { }
-        lowerHandleNonce = i;
-        if (i != 0) {
-            recoveryBlockNumber = handleNoncesToBlockNums[i-1];
+        for (i = lowerHandleNonce; i <= upperHandleNonce && handleNoncesToBlockNums[i] > 0; i++) {
+            recoveryBlockNumber = handleNoncesToBlockNums[i];
+            delete handleNoncesToBlockNums[i];
+            delete closedValueTransferVotes[i];
```
Comment (jeonghonah) [inline contracts/bridge/BridgeTransferCommon.sol]: Since `closedValueTransferVotes` is related the multi operator, how about moving this code also?

Comment (KimKyungup) [reply contracts/bridge/BridgeTransferCommon.sol]: I think this `closedValueTransferVotes` can prevent duplicated voting. So I don't have no idea about another position.

File: contracts/bridge/BridgeOperator.sol [hunk 7]
```diff
@@ -23,9 +23,14 @@ import "../externals/openzeppelin-solidity/contracts/ownership/Ownable.sol";
...
 contract BridgeOperator is Ownable {
+    struct VotesData {
+        mapping(bytes32 => mapping(address => bool)) voted; // <sha3(type, args, nonce), <operator, vote>>
+        mapping(bytes32 => uint8) voteCounts; // <sha3(type, args, nonce)>
+    }
+
     mapping(address => bool) public operators;
-    mapping(bytes32 => mapping(address => bool)) public votes; // <sha3(type, args, nonce), <operator, vote>>
-    mapping(bytes32 => uint8) public votesCounts; // <sha3(type, args, nonce)>
+    mapping(uint8 => mapping (uint64 => VotesData) ) private votes;
```
Comment (nogang) [inline contracts/bridge/BridgeOperator.sol]:
```suggestion
    mapping(uint8 => mapping (uint64 => VotesData)) private votes;
```

Comment (KimKyungup) [reply contracts/bridge/BridgeOperator.sol]: fixed. :)

**Comment by KimKyungup** [comment]
@jeonghonah @nogang PTAL. :)

**Comment by aidan-kwon** [comment]
It seems that only one vote can be processed at a time in your design. 
If you intended and want to remove all past voting data, 
how about changing types of variables in `VotesData` like the below.
```
struct VotesDate {
	bytes32 voteKey; 
	uint8 voteCount; 
	mapping(address => bool) voted;
}
```

**Comment by KimKyungup** [comment]
> @KimKyungup It seems that only one vote can be processed at a time in your design.
> If you intended and want to remove all past voting data,
> how about changing types of variables in `VotesData` like the below.
> 
> ```
> struct VotesDate {
> 	bytes32 voteKey; 
> 	uint8 voteCount; 
> 	mapping(address => bool) voted;
> }
> ```

@aidan-kwon For each nonce, operators can vote different voteKey. Through this code I would like to remove all vote data at each nonce. So I wrote `VoteData` like it.

**Comment by aidan-kwon** [comment]
> > @KimKyungup It seems that only one vote can be processed at a time in your design.
> > If you intended and want to remove all past voting data,
> > how about changing types of variables in `VotesData` like the below.
> > ```
> > struct VotesDate {
> > 	bytes32 voteKey; 
> > 	uint8 voteCount; 
> > 	mapping(address => bool) voted;
> > }
> > ```
> 
> @aidan-kwon For each nonce, operators can vote different voteKey. Through this code I would like to remove all vote data at each nonce. So I wrote `VoteData` like it.

Thank you. I understand that. Then, how about 
```
struct VotesDate {
	mapping(address => bytes32) voted;
	mapping(bytes32 => unint8) voteCount;
}
```

File: contracts/bridge/BridgeOperator.sol [hunk 8]
```diff
@@ -23,9 +23,14 @@ import "../externals/openzeppelin-solidity/contracts/ownership/Ownable.sol";
...
 contract BridgeOperator is Ownable {
+    struct VotesData {
+        mapping(bytes32 => mapping(address => bool)) voted; // <sha3(type, args, nonce), <operator, vote>>
+        mapping(bytes32 => uint8) voteCounts; // <sha3(type, args, nonce)>
+    }
+
     mapping(address => bool) public operators;
-    mapping(bytes32 => mapping(address => bool)) public votes; // <sha3(type, args, nonce), <operator, vote>>
-    mapping(bytes32 => uint8) public votesCounts; // <sha3(type, args, nonce)>
+    mapping(uint8 => mapping (uint64 => VotesData)) private votes;
```
Comment (aidan-kwon) [inline contracts/bridge/BridgeOperator.sol]:
```suggestion
    mapping(uint8 => mapping (uint64 => VotesData)) private votes; // <voteType>, <voteType' nonce, voteData>
```

File: contracts/bridge/BridgeOperator.sol [hunk 9]
```diff
@@ -23,9 +23,14 @@ import "../externals/openzeppelin-solidity/contracts/ownership/Ownable.sol";
...
+
```
Comment (aidan-kwon) [inline contracts/bridge/BridgeOperator.sol]:
```suggestion
```

**Comment by KimKyungup** [comment]
@aidan-kwon @jeonghonah @nogang PTAL. :)

**Comment by KimKyungup** [comment]
@aidan-kwon @jeonghonah @nogang PTAL again. I modified for perfect deleting. :)

File: contracts/bridge/BridgeOperator.sol [hunk 10]
```diff
@@ -23,11 +23,18 @@ import "../externals/openzeppelin-solidity/contracts/ownership/Ownable.sol";
 
 
 contract BridgeOperator is Ownable {
+    struct VotesData {
+        address[] voters;   // voter list for deleting voted map
```
Comment (jeonghonah) [inline contracts/bridge/BridgeOperator.sol]: How about using operators?

File: contracts/bridge/BridgeOperator.sol [hunk 11]
```diff
@@ -23,11 +23,18 @@ import "../externals/openzeppelin-solidity/contracts/ownership/Ownable.sol";
...
+        mapping(address => bytes32) voted; // <operator, sha3(type, args, nonce)>
+
+        bytes32[] voteKeys; // voteKey list for deleting voteCounts map
```
Comment (jeonghonah) [inline contracts/bridge/BridgeOperator.sol]: How about find keys by using operators?

**Pull request merged**